hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java

```diff
@@ -0,0 +1,534 @@
+        if (!listKeysMode &&
+            !StringUtils.startsWith(startKey, keyName) &&
+            !OzoneFSUtils.isImmediateChild(keyName, startKey)) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("StartKey {} is not an immediate child of keyName {}. " +
+                "Returns empty list", startKey, keyName);
+          }
+          return new ArrayList<>();
+        }
+      } else {
+        keyName = OzoneFSUtils.getParentDir(startKey);
+        prefixKey = keyName;
+        args = args.toBuilder()
+            .setKeyName(keyName)
+            .setSortDatanodesInPipeline(false)
+            .build();
+      }
+    }
+
+    OzoneFileStatus fileStatus =
+        getStatusHelper.apply(args, clientAddress, listKeysMode);
+
+    String dbPrefixKey;
+    if (fileStatus == null) {
+      // if the file status is null, prefix is a not a valid filesystem path
+      // this should only work in list keys mode.
+      // fetch the db key based on the prefix path.
+      dbPrefixKey = getDbKey(keyName, args, omBucketInfo);
+      prefixKey = OzoneFSUtils.getParentDir(keyName);
+    } else {
+      // If the keyname is a file just return one entry
+      if (fileStatus.isFile()) {
+        return Collections.singletonList(fileStatus);
+      }
+
+      // fetch the db key based on parent prefix id.
+      long id = getId(fileStatus, omBucketInfo);
+      dbPrefixKey = metadataManager.getOzonePathKey(id, "");
+    }
+
+    // Determine startKeyPrefix for DB iteration
+    String startKeyPrefix =
+        Strings.isNullOrEmpty(startKey) ? "" :
+            getDbKey(startKey, args, omBucketInfo);
+
+    TreeMap<String, OzoneFileStatus> map = new TreeMap<>();
+
+    BucketLayout bucketLayout = omBucketInfo.getBucketLayout();
+
+    // fetch the sorted output using a min heap iterator where
+    // every remove from the heap will give the smallest entry.
+    try (MinHeapIterator heapIterator = new MinHeapIterator(metadataManager,
+        dbPrefixKey, bucketLayout, startKeyPrefix, volumeName, bucketName)) {
+
+      while (map.size() < numEntries && heapIterator.hasNext()) {
+        HeapEntry entry = heapIterator.next();
+        OzoneFileStatus status = entry.getStatus(prefixKey,
+            scmBlockSize, volumeName, bucketName);
+        map.put(entry.key, status);
+      }
+    }
+
+    return map.values();
+  }
+
+  private String getDbKey(String key, OmKeyArgs args,
+                          OmBucketInfo omBucketInfo) throws IOException {
+    long startKeyParentId;
+    OzoneFileStatus fileStatusInfo = getFileStatus(key, args);
+    if (fileStatusInfo == null || fileStatusInfo.isFile()) {
+      String parent = OzoneFSUtils.getParentDir(key);
+
+      fileStatusInfo = getFileStatus(parent, args);
+
+      // the keyname is not a valid filesystem path.
+      // determine the parent prefix by fetching the
+      // prefix id of the parent directory
+      Preconditions.checkNotNull(fileStatusInfo);
```

Comment:
checkNotNull can throw NPE, which is a RuntimeException and can leads to retries. Instead of that, can you throw OMException explicitly.

```
2022-05-26 10:37:31,235 [Time-limited test] INFO  retry.RetryInvocationHandler (RetryInvocationHandler.java:log(411)) - com.google.protobuf.ServiceException: org.apache.hadoop.ipc.RemoteException(java.lang.NullPointerException): java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:889)
	at org.apache.hadoop.ozone.om.OzoneListStatusHelper.getDbKey(OzoneListStatusHelper.java:)
```


```
  public static <T> T checkNotNull(@CheckForNull T reference) {
    if (reference == null) {
      throw new NullPointerException();
    } else {
      return reference;
    }
  }
```